[580] Snap to grid on center for not-resizable node

peransin-obeo opened the pull request:
When moving or creating fixed size nodes, the center of the node snap to center instead of its side.

For those nodes, creating edges will be aligned to grid.

## Steps to tests
- Use provided VSM to create elements
[noderefresh.zip](https://github.com/user-attachments/files/19623490/noderefresh.zip)
- Create a diagram with snap-on-grid, displayed grid, and grid spacing = 40
- Creates elements: Instance 
- "Instance" must be centered vertically on creation and when moved
- "Property" and "Association" on "Class" must be centered on creation and when moved

peransin-obeo committed:
[580] Anchor nodes by its center on grid when not resizable

When an axis is not resizable, its center is set on Grid point to ease
edge alignment.

This applies on drag-and-drop and when creating a node.

Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/580
Signed-off-by: Nicolas PERANSIN <nicolas.peransin@obeo.fr>

lredor commented on plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/internal/refresh/NodePositionHelper.java:
```diff
@@ -61,7 +61,7 @@ public class NodePositionHelper {
      */
     public NodePositionHelper(boolean snapToGrid, int gridSpacing) {
-        this.snapToGrid = snapToGrid;
+        this.snapToGrid = snapToGrid && gridSpacing > 0;
```
I think that this change must be documented in the javadoc.

lredor commented on plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/internal/refresh/NodePositionHelper.java:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    /**
+     * Gets the shift for a node to be centered on a point if not resizable.
```
```suggestion
     * Gets the shift for a node to be centered on a point of the grid, if the node is not resizable.
```

lredor reviewed (changes requested):
Only the commit "[580] Anchor nodes by its center on grid when not resizable" must be in this PR. This PR should follow the PR #581. In the current configuration the 3 commits from PR #581 also appear in PR #582. In this review, I consider that these 3 commits are the same (not reviewed).

peransin-obeo commented on plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/ruler/SiriusSnapToGrid.java:
```diff
@@ -0,0 +1,65 @@
+                && request.getEditParts().get(0) instanceof ShapeNodeEditPart ep // nodes
+                && ep.getModel() instanceof Node node && node.getLayoutConstraint() instanceof Size size) {
+
+            // It is easier to shift the origin than to alterate the result
```
Done.

peransin-obeo committed:
[580] Update Tests related to Node position

NodeCreationWithSnapToGridTest is impacted as created nodes are
not resizable.
From previous commit, such nodes are shifted to align its center with a
Grid point.

Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/580
Signed-off-by: Nicolas PERANSIN <nicolas.peransin@obeo.fr>

peransin-obeo commented:
Draft: updating release note.

lredor commented on plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:
```diff
@@ -6,7 +6,11 @@ This document contains the release notes for recent major releases of Sirius. Se
-Nothing.
+h3. User-Visible Changes
+
+* <span class="label label-info">Modified</span> Edition using the Grid is enhanced. 
```
```suggestion
* <span class="label label-info">Modified</span> Edition using the Grid has been improved.
```

lredor commented on plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:
```diff
@@ -6,7 +6,11 @@ This document contains the release notes for recent major releases of Sirius. Se
+h3. User-Visible Changes
+
+* <span class="label label-info">Modified</span> Edition using the Grid is enhanced. 
+** When creating a node with active "Snap To Grid", the size is adapted to the Grid Spacing so the resizable sides fit on it. This behavior results in having corners of the node matching the grid. Currently, it only applies if the node mapping has size indications to provide a preferred size.
```
```suggestion
** When creating a node with "Snap to Grid" enabled, its size is adjusted to the grid spacing so that its resizable sides align with the grid. This ensures that the node’s corners match the grid intersections. Currently, this only applies when the node mapping provides size hints to define a preferred size.
```

lredor replied:
Ideally, this first part should only have concerned the previous PR #581. Something to keep in mind (and to update) if only the first PR gets merged.

lredor commented on plugins/org.eclipse.sirius.doc/doc/Release_Notes.textile:
```diff
@@ -6,7 +6,11 @@ This document contains the release notes for recent major releases of Sirius. Se
+
+* <span class="label label-info">Modified</span> Edition using the Grid is enhanced. 
+** When creating a node with active "Snap To Grid", the size is adapted to the Grid Spacing so the resizable sides fit on it. This behavior results in having corners of the node matching the grid. Currently, it only applies if the node mapping has size indications to provide a preferred size.
+** When creating or moving a node with active "Snap To Grid", the position is adapted to the Grid Spacing so fixed-size sides are centered. This behavior results in a better alignment and centering of edges and for most of border and image-based nodes.
```
```suggestion
** When creating or moving a node with "Snap to Grid" enabled, its position is adjusted to the grid spacing so that fixed-size sides are centered. This results in better alignment and centering of edges, as well as for most border and image-based nodes.
```

peransin-obeo committed:
[580] Update Release Note for Grid usage

Detail influence of Grid on node positions when moved or created.

lredor reviewed (approved):
It seems good for me now. A validation of tests from https://ci.eclipse.org/sirius/job/sirius-pr-check/view/change-requests/job/PR-582/ is nevertheless necessary. I don't understand why SwtBot tests part fails.
